gcloud/src/auth.rs

```diff
@@ -0,0 +1,98 @@
+use std::error::Error;
+use std::io;
+use std::path::Path;
+
+use hyper::Client;
+
+use snafu::{ResultExt, Snafu};
+
+use yup_oauth2::{GetToken, ServiceAccountAccess, Token};
+
+use crate::https;
+
+/// Possible auth errors.
+#[derive(Debug, Snafu)]
+#[allow(missing_docs)] // Otherwise, cargo will ask to document each field of each error, which is a bit overkill.
+pub enum AuthError {
+    /// Error returned when the service account key file couldn't be read
+    /// by the authentication provider.
+    #[snafu(display("Failed to read service account key: {}", source))]
+    FailedToReadServiceAccountKey { source: io::Error },
+}
+
+type Result<T> = std::result::Result<T, AuthError>;
+
+/// Provides google cloud authentication with a simple API.
+///
+/// It requires less boilerplate than the bare `yup_oauth2` provider, and it implements
+/// the `oauth::GetToken` trait, which means it can be used directly with Pubsub & Datastore clients.
+pub struct AuthProvider {
+    access_manager: ServiceAccountAccess<Client>,
+}
+
+impl AuthProvider {
+    /// Create an auth provider from a JSON file on disk.
+    ///
+    /// # Errors
+    /// Returns an AuthError if the file cannot be read or if it doesn't contain
+    /// valid JSON.
+    pub fn from_json_file<P: AsRef<Path>>(path: &P) -> Result<AuthProvider> {
+        // This is needed because `yup_oauth2` doesn't support `std::Path`...
+        let path_str = String::from(path.as_ref().to_string_lossy());
+
+        let secret = yup_oauth2::service_account_key_from_file(&path_str)
+            .context(FailedToReadServiceAccountKey)?;
+
+        let access_manager = ServiceAccountAccess::new(secret, https::new_tls_client());
+
+        Ok(AuthProvider { access_manager })
+    }
+}
+
+impl GetToken for AuthProvider {
+    fn token<'b, I, T>(&mut self, scopes: I) -> std::result::Result<Token, Box<dyn Error>>
```

Comment:
Corrige moi si je suis dans le champ. Mais, en gros, token prend deux types génériques où `T` doit implémenter les traits AsRef<str> et Ord en plus de respecter le named lifetime spécifié par 'b et `I` doit implémenter le trait IntoIterator où les Items sont de type T et respectent le lifetime 'b?

Reply:
Spot on. (very nice compréhension c'est vraiment pas un exemple évident). En gros la fonction reçoit un objet quelconque (`I`) qui peut être transformé en un itérateur de `T`s. La spécification de lifetime indique au compilateur que les items retournés par IntoIterator doivent vivre au moins aussi longtemps que l'objet de type `I` qui crée l'iterateur.